Improve the "copy to clipboard" functionality

m-vo opened the pull request:
Supersedes https://github.com/contao/contao/pull/5612

This adds add flash message when copying things to the clipboard as suggested in https://github.com/contao/contao/pull/5612#issuecomment-2720151593.

We can now easily display messages in JS with the new `Message` module, that works together with the main message outlet:

```js
import { Message } from '../modules/message';

// …

Message.info('Something worked.');
Message.error('Oops, there was an error.');
```

Messages are pre-rendered with Twig - so this would even work with adjusted templates.

m-vo committed:
add message prototypes system, so that we can render messages from JS

m-vo committed:
add message module to ease working with messages

m-vo committed:
improve copy to clipboard functionality

m-vo committed:
build assets

zoglo commented on core-bundle/contao/languages/en/tl_preview_link.xlf:
```diff
@@ -50,6 +50,9 @@
         <source>Copy the preview link ID %s to your clipboard.</source>
       </trans-unit>
+      <trans-unit id="tl_preview_link.copied_message">
+        <source>The link was copied to the clipboard.</source>
```
```suggestion
        <source>Copied to clipboard!</source>
```

The default sounds way better, don't you think?

m-vo replied:
If we go that route, the key should probably be in `MSC` then.

m-vo replied:
See 8d57652b51c1a6660a8ac0d08af33171f993d7d1

zoglo commented on core-bundle/assets/controllers/message-outlet-controller.js:
```diff
@@ -13,6 +13,21 @@ export default class extends Controller {
+                this.element.append(document.createRange().createContextualFragment(html));
+
+                return;
+            }
```
```suggestion
            if (target.dataset.type === type) {
                this.element.append(document.createRange().createContextualFragment(target.getHTML().replace('{{message}}', message)));
                return;
            }
```
Could eliminate that const, right?

fritzmg replied:
Too long for single line imho.

m-vo replied:
This will get compiled anyways.

m-vo committed:
use a default message

aschempp commented:
Any particular reason to remove the fallback? Maybe we should remove the operation if `navigator.clipboard` is not supported (or the back end is not accessed in a secure context (HTTPS))?

m-vo commented:
> Any particular reason to remove the fallback? 

See https://github.com/contao/contao/pull/5612#discussion_r1064123322

> Maybe we should remove the operation if [...] the back end is not accessed in a secure context (HTTPS)?

Why should we do this? Could you elaborate a bit?

aschempp commented:
Because `navigator.clipboard` only works in secure contexts (HTTPS). I assume the button would not do anything if the action fails?

m-vo committed:
Merge remote-tracking branch 'origin/5.x' into feature/copy-to-clipboard-2.0

m-vo committed:
CS

m-vo committed:
handle errors

m-vo committed:
build assets

m-vo committed:
add an icon to clipboard links

m-vo committed:
build assets

m-vo commented:
I implemented error handling in e1a224c14b80ca4f2d83479ad8fc4410a0228fa0 and added an icon to all "clipboard links" (`a[data-controller="contao--clipboard"]`), so they get more obvious:

![image](https://github.com/user-attachments/assets/2eeb4be5-1f0e-4d21-b763-d83b40f21862)

![image](https://github.com/user-attachments/assets/7eff4cac-9272-4302-ad0d-7583b6a2e673)

aschempp commented on core-bundle/assets/controllers/clipboard-controller.js:
```diff
@@ -1,25 +1,26 @@
+    async write() {
+        if (!navigator.clipboard) {
+            if (window.console) {
+                console.error('The clipboard API is not available - make sure you are using a secure context (https).');
```
shouldn't we render an error message popup instead? A regular user won't see that when clicking a button.

m-vo replied:
IMHO not. Having no secure context is always a configuration error/should only ever happen during development. So this is no user message.

(Btw, in Firefox and Chrome (at least on Linux) the clipboard API even worked for local non-https domains.)

leofeyer replied:
Then why log an error message at all?

m-vo replied:
IMHO still valuable, because you then get an idea why nothing is copied.

leofeyer committed:
Merge branch '5.x' into feature/copy-to-clipboard-2.0

leofeyer commented on core-bundle/assets/styles/component/clipboard.pcss:
```diff
@@ -0,0 +1,18 @@
+        margin-right: 2px;
+        margin-bottom: 2px;
+        vertical-align: middle;
+        background: url("../../../contao/themes/flexible/icons/clipboard.svg") no-repeat center center;
```
Unfortunately, the icon you added says "remove from clipboard" or "close clipboard" instead of "copy to clipboard".

m-vo replied:
That's the icon we already had. 🤔 Should we exchange it? Or in which context is it used?

leofeyer replied:
As I said, we use it for "close clipboard":

<img width="306" alt="" src="https://github.com/user-attachments/assets/92739fbb-255b-4cf9-be6d-6a195dc4d47c" />

m-vo replied:
Ah, now I got you. 🤦‍♂️ True. Should we introduce a new one or reuse one of the "arrow" ones? Wdyt?

leofeyer replied:
We need a new one with a plus I guess.

fritzmg replied:
Which one of these should we use?

* [`clipboard-plus`](https://lucide.dev/icons/clipboard-plus) ![clipboard-plus](https://github.com/user-attachments/assets/1392a1d5-b1ab-47a7-a7e9-7433063b4936)
* [`clipboard-copy`](https://lucide.dev/icons/clipboard-copy) ![clipboard-copy](https://github.com/user-attachments/assets/04ebaff5-fae6-4398-9e47-370ebcae42e9)
* [`clipboard-list`](https://lucide.dev/icons/clipboard-list) ![clipboard-list](https://github.com/user-attachments/assets/807f8857-6911-40c1-8f42-40353d7a0382)

zoglo replied:
`clipboard-plus` looks too close to "close".
`clopboard-copy` seems more intuitive imo

fritzmg replied:
My first intuition is `clipboard-plus` within Contao as `clipboard-copy` looks too close to the paste into/after icons. But I am also not sure.

leofeyer replied:
`clipboard-plus` is the only one of the three that says "add", so we should use this.

m-vo replied:
Now using the `clipboard-plus` icon in 4a40e2cc601cc66672432e71e8027b92cd79a284.

m-vo committed:
exchange icon

m-vo committed:
Merge remote-tracking branch 'origin/5.x' into feature/copy-to-clipboard-2.0

m-vo committed:
build assets

m-vo commented:
This one is also ready, IMHO. Conflicts are resolved again.

leofeyer committed:
Merge branch '5.x' into feature/copy-to-clipboard-2.0

leofeyer committed:
Merge branch '5.x' into feature/copy-to-clipboard-2.0

leofeyer committed:
CS

m-vo committed:
use "export const Message" and run linter

leofeyer committed:
Merge branch '5.x' into feature/copy-to-clipboard-2.0

leofeyer committed:
Rebuild the assets
